File: Serenity.Script.UI/SlickGrid/Formatters/EnumTypeRegistry.cs
using System;
using System.Collections.Generic;

namespace Serenity
{
    public static class EnumTypeRegistry
    {
        private static JsDictionary<string, Type> knownTypes;

        public static Type Get(string key)
        {
            if (knownTypes == null)
            {
                knownTypes = new JsDictionary<string, Type>();
                foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
                {
                    foreach (var type in assembly.GetTypes())
                    {
                        if (type.IsEnum)
                        {
                            var fullName = type.FullName;
                            knownTypes[fullName] = type;

                            var enumKeyAttr = type.GetCustomAttributes(typeof(EnumKeyAttribute), false);
                            if (enumKeyAttr != null && enumKeyAttr.Length > 0)
                                knownTypes[((EnumKeyAttribute)enumKeyAttr[0]).Value] = type;

                            foreach (var k in Q.Config.RootNamespaces)
                            {
                                if (fullName.StartsWith(k + "."))
                                {
                                    knownTypes[fullName.Substr(k.Length + 1)] = type;
                                }
                            }
                        }
                    }
                }
            }

            if (!knownTypes.ContainsKey(key))
                throw new Exception(String.Format("Can't find {0} enum type! If you have recently defined this enum type in server side code, " + 
                    "make sure your project builds successfully and transform T4 templates. Also make sure that enum is under your project root namespace, " + 
                    "and your namespace parts starts with capital letters, e.g. MyProject.Pascal.Cased namespace", key));

            return knownTypes[key];
        }
    }
}

File: src/Serenity.Scripts/CoreLib/Types/EnumTypeRegistry.ts
import { EnumKeyAttribute } from "../Decorators";
import { Config, Exception, format, getAttributes, getTypeFullName, getTypes, isEnum, notifyError, startsWith } from "../q";

export namespace EnumTypeRegistry {

    let knownTypes: { [key: string]: Function }

    export function tryGet(key: string): Function {

        if (knownTypes == null) {
            knownTypes = {};
            for (var type of getTypes()) {
                if (isEnum(type)) {
                    var fullName = getTypeFullName(type);
                    knownTypes[fullName] = type;
                    var enumKeyAttr = getAttributes(type, EnumKeyAttribute, false);
                    if (enumKeyAttr != null && enumKeyAttr.length > 0) {
                        knownTypes[enumKeyAttr[0].value] = type;
                    }

                    for (var k of Config.rootNamespaces) {
                        if (startsWith(fullName, k + '.')) {
                            knownTypes[fullName.substr(k.length + 1)] = type;
                        }
                    }
                }
            }
        }

        if (knownTypes[key] == null)
            return null;

        return knownTypes[key];
    }

    export function get(key: string): Function {
        var type = EnumTypeRegistry.tryGet(key);
        if (type == null) {
            var message = format("Can't find {0} enum type! If you have recently defined this enum type " + 
                "in server side code, make sure your project builds successfully and transform T4 templates. " +
                "Also make sure that enum is under your project root namespace, and your namespace parts starts " + 
                "with capital letters, e.g.MyProject.Pascal.Cased namespace", key);

            notifyError(message, '', null);
            throw new Exception(message);
        }

        return type;
    }
}